python/mage/tgn/definitions/tgn.py

```diff
@@ -0,0 +1,1213 @@
+        # m_ij = (s_i, s_j, delta t, e_ij)
+        # delta t is only one number :)
+        self.edge_raw_message_dimension = (
+            2 * self.memory_dimension + 1 + num_edge_features
+        )
+
+        # dimension of raw message for node
+        # m_i = (s_i, t, v_i)
+        self.node_raw_message_dimension = self.memory_dimension + 1 + num_node_features
+
+        self.raw_message_store = RawMessageStore(
+            edge_raw_message_dimension=self.edge_raw_message_dimension,
+            node_raw_message_dimension=self.node_raw_message_dimension,
+        )
+
+        MessageFunctionEdge = get_message_function_type(edge_message_function_type)
+
+        # if edge function is identity, when identity function is applied, it will result with
+        # vector of greater dimension then when identity is applied to node raw messsage, so node function must be MLP
+        # if edge is MLP, node also will be MLP
+        MessageFunctionNode = MessageFunctionMLP
+
+        self.message_dimension = message_dimension
+
+        if edge_message_function_type == MessageFunctionType.Identity:
+            # We set message dimension (dimension after message function is applied to raw message) to
+            # dimension of raw message of edge interaction event, since message dimension will be
+            # just dimension of concatenated vectors in edge interaction event.
+            # Also, MLP in node event will then produce vector of same dimension and we can then aggregate these
+            # two vectors, with MEAN or LAST
+            self.message_dimension = self.edge_raw_message_dimension
+
+        self.edge_message_function = MessageFunctionEdge(
+            message_dimension=self.message_dimension,
+            raw_message_dimension=self.edge_raw_message_dimension,
+        )
+
+        self.node_message_function = MessageFunctionNode(
+            message_dimension=self.message_dimension,
+            raw_message_dimension=self.node_raw_message_dimension,
+        )
+
+        MessageAggregator = get_message_aggregator_type(message_aggregator_type)
+
+        self.message_aggregator = MessageAggregator()
+
+        self.memory = Memory(memory_dimension=memory_dimension)
+
+        MemoryUpdaterType = get_memory_updater_type(memory_updater_type)
+
+        self.memory_updater = MemoryUpdaterType(
+            memory_dimension=self.memory_dimension,
+            message_dimension=self.message_dimension,
+        )
+
+        self.time_encoder = TimeEncoder(out_dimension=self.time_dimension)
+
+    def memory_detach_tensor_grads(self):
+        self.memory.detach_tensor_grads()
+        # self.raw_message_store.detach_grads()
+
+    def forward(
+        self,
+        data: Tuple[
+            np.ndarray,
+            np.ndarray,
+            np.ndarray,
+            np.ndarray,
+            Dict[int, torch.Tensor],
+            Dict[int, torch.Tensor],
+        ],
+    ):
+        raise Exception("Not implemented")
```

Comment:
What should happen here?

Reply:
Nothing, I implemented this in `TGNEdgesSelfSupervised`, `TGNSupervised`, so I put all logic there. This is not intended to be used, so don't know what else to do.